src/UpgradeRule/PHP/Visitor/SymbolContextVisitor.php

```diff
@@ -0,0 +1,167 @@
+<?php
+
+namespace SilverStripe\Upgrader\UpgradeRule\PHP\Visitor;
+
+use PhpParser\Node;
+use PhpParser\NodeVisitor;
+use PhpParser\Node\Stmt\Use_;
+use PhpParser\Node\Stmt\Class_;
+use PhpParser\Node\Stmt\ClassMethod;
+use PhpParser\Node\Stmt\Namespace_;
+use PhpParser\Node\Stmt\PropertyProperty;
+use PhpParser\Node\Expr\New_;
+use PhpParser\Node\Expr\MethodCall;
+use PhpParser\Node\Expr\StaticCall;
+use PhpParser\Node\Expr\PropertyFetch;
+use PhpParser\Node\Expr\StaticPropertyFetch;
+
+/**
+ * Accumulates symbols found in a class that might be meaningful
+ * in a fuzzy use matching. For example, if the file contains
+ * a "use SilverStripe\Forms\GridField\GridField" statement,
+ * and later references "$myField->someRemovedMethod()",
+ * we can infer that a rule with "GridField->someRemovedMethod()"
+ * likely applies.
+ *
+ * The context can change for each node in an AST, for example a static method
+ * call to "MyClass::myMethod()" will set $staticInvokedClass,
+ * and invoke it after finishing processing of the method call node.
+ *
+ * Since the PHP syntax parser works sequentially, it will only determine
+ * context that's defined before the symbol in question is used.
+ *
+ * Since the PHP syntax parser doesn't actually understand symbol relationships,
+ * and doesn't resolve dependencies and hierarchies, this isn't a very reliable
+ * methodology, and should only be used for warnings rather than code rewrites.
+ *
+ * @package SilverStripe\Upgrader\Util
+ */
+class SymbolContextVisitor implements NodeVisitor {
```

Comment:
So is Symbol Context a "type hint" or "inferred type"? I'd find a name based on either of those phrases to make more sense. "type inferencing" is essentially what you're doing here.

Also - this visitor is made more complex by having to parse its own namespace / use data. There's another visitor that turns everything into FQCNs and I'd suggest you rely on that.

Note also that this visitor could be make a lot more powerful in a future PR. ;-)

Reply:
https://github.com/silverstripe/silverstripe-upgrader/blob/master/src/UpgradeRule/PHP/Visitor/ClassQualifierVisitor.php is the visitor in question

Reply:
Alternatively you could make this this visitor a subclass of NameResolver as ClassQualifierVisitor does

Reply:
> So is Symbol Context a "type hint" or "inferred type"? 

That's not what the visitor does though. It's collecting context. You could call those hints which are then later used as the basis for type inference, but I think it'd be confusing to call the visitor something like that. E.g. its storing all `use`statements on a method. A method doesn't have a "type" in the sense that we care about, the class instance this method is called on does.

> Alternatively you could make this this visitor a subclass of NameResolver as ClassQualifierVisitor does

Good call - once I've understood what `NameResolver` does (it's not documented in the upstream package), it proved useful hehe.